_posts/2016-10-06-database-testing-with-tsqlt.md

```diff
@@ -0,0 +1,87 @@
+We first need to get the [download](http://tsqlt.org/downloads/) from the official website. 
+
+The tSQLt [open source](https://github.com/tSQLt-org/tSQLt) framework allows us to run our tests within transactions which makes our tests independant and takes care of cleaning up after it runs. The framework also gives us the ability to isolate our tests through fake tables and stored procedure (SP) spies.
+
+##Running the test sample
+
+Once you have SQL Server Express up and running, and MSSQL Management Studio installed, open the latter and connect to your server.
+
+{% img /assets/img/custom/blog/db-testing/db-testing-1.png "" %}
+
+The tSQLt team provides a quick start example database and a set of tests to help developers get started quickly on how the framework works.
+Let’s try the example together.
+First unzip the tSQLt folder you downloaded, and make sure that CLRs are enabled on your development server. Enabling [CLR integration](https://msdn.microsoft.com/en-us/library/ms254498v=vs.110.aspx) allows us to run managed code, such as C#, on our SQL Server database.
+
+{% img /assets/img/custom/blog/db-testing/db-testing-2.png "" %}
+
+Open the Example.sql file that you will find within the unzipped folder, and execute it in your database server. This will create a test database named `tSQLt_Example`.
+
+Once your database is created, open a new query window on that DB and run the following command:
+`EXEC tSQLt.RunAll`
+This is a command to run all the tests. You should see a failing test in the result screen:
+
+{% img /assets/img/custom/blog/db-testing/db-testing-3.png "" %}
+
+Test case number 11, named test ready for experimentation if 2 particles, is failing. Let’s open the file and see what’s wrong with it; we will need to navigate to the stored procedures folder to find the test.
+
+{% img /assets/img/custom/blog/db-testing/db-testing-4.png "" %}
+
+{% img /assets/img/custom/blog/db-testing/db-testing-5.png "" %}
+
+The failing test is calling `IsExperimentReady` and checks that it returns 1 when the particle table has two particles.
+Looking at the function, it seems that it is not counting the number of rows in the Particle table, and it is not doing the right checks.
+
+{% img /assets/img/custom/blog/db-testing/db-testing-6.png "" %}
+
+Once we’ve solved the issue, we can run the tests again and see that they are all passing now.
+
+###What is this test doing and how is the fake table working?
+
+Tests in the tSQLt framework are run within transactions. The function we are testing relies on the Particle table to determine whether the system is ready for experimentation. Looking at our test here, we can see that we are creating a fake table of the `Accelerator.Particle` in which we will insert two rows.
+
+If we have closer at look at the `tSQLt.FakeTable` SP, it is renaming the original table we are faking, and copying its structure into a new table of the original name without applying any of the constraints. 

```

Comment:
If we have closer **look** at the `tSQLt.FakeTable` stored procedure, **we can see** it is renaming ...